pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,210 @@
+// It is called when the IP identity cache changes and the IP is deleted from the mapping
+// It is also called when the IP is upserted with a new identity
+// It removes the prefix from the prefixLengths map
+// It is called with the identityToIpMutex lock held
+func (s *FQDNDataServer) deleteFromIdentityToIPLocked(identity *ipcache.Identity, cidr string, prefix netip.Prefix) error {
+	if identity == nil {
+		return fmt.Errorf("identity is nil")
+	}
+
+	if ips, ok := s.currentIdentityToIp[identity.ID]; ok {
+		newIps := slices.DeleteFunc(ips, func(existing string) bool {
+			if existing == cidr {
+				s.prefixLengths.Delete([]netip.Prefix{prefix})
+				return true
+			}
+			return false
+		})
+		if len(newIps) == 0 {
+			delete(s.currentIdentityToIp, identity.ID)
+		} else {
+			s.currentIdentityToIp[identity.ID] = newIps
+		}
+	}
+	return nil
+}
+
+// UpdatePolicyRules updates the current state of the DNS rules with the given policies and sends the current state of the DNS rules to the client
+// This method is called:
+// 1. when the DNS rules are updated during the endpoint regeneration, we store the state of the DNS rules with flag rulesUpdate as true
+// 2. when the client subscribes to DNS policies, we send the current state of the DNS rules to the client(flag rulesUpdate as false)
+// 3. when the IP identity cache changes, we update the current state of the identity to IP mapping and send the current state of the DNS rules to
+// the client(flag rulesUpdate as false)
+// Note: this method is left empty on purpose and will be updated with the actual implementation in the future PRs for the standalone DNS proxy
+func (s *FQDNDataServer) UpdatePolicyRules(policies map[identity.NumericIdentity]policy.SelectorPolicy, rulesUpdate bool) error {
+	return nil
+}
+
+// UpdateMappingRequest updates the FQDN mapping with the given data
+// SDP sends the fqdn mapping to cilium agent
+// Steps to update the mapping:
+// 1. Get the endpoint from the IP
+// 2. If the endpoint is not found, return an error
+// 3. If the IPs are not empty, update the cilium agent with the mapping
+// Note: this method is left empty on purpose and will be updated with the actual implementation in the future PRs for the standalone DNS proxy
+func (s *FQDNDataServer) UpdateMappingRequest(ctx context.Context, mappings *pb.FQDNMapping) (*pb.UpdateMappingResponse, error) {
+	return &pb.UpdateMappingResponse{
+		Response: pb.ResponseCode_RESPONSE_CODE_NO_ERROR,
+	}, nil
+}
+
+// ListenAndServe starts the Standalone DNS Proxy gRPC server on the given port
+func (s *FQDNDataServer) ListenAndServe() error {
+	address := fmt.Sprintf("localhost:%d", s.port)
+	s.log.Info("Starting Standalone DNS Proxy server on", logfields.Address, address)
+	lis, err := net.Listen("tcp", address)
+	if err != nil {
+		s.log.Error("Failed to listen", logfields.Error, err)
+		return err
+	}
+	grpcServer := grpc.NewServer(grpc.KeepaliveEnforcementPolicy(kaep), grpc.KeepaliveParams(kasp))
```

Comment:
You can create, assign and register this server as part of the constructor, this intentionally doesn't need the listener (or any runtime thing) yet. That also makes a race with `Stop` less likely.